src/lib/tls/asio/asio_stream_core.h

```diff
@@ -0,0 +1,120 @@
+/*
+* TLS Stream Helper
+* (C) 2018-2019 Jack Lloyd
+*     2018-2019 Hannes Rantzsch, Tim Oesterreich, Rene Meusel
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_ASIO_STREAM_CORE_H_
+#define BOTAN_ASIO_STREAM_CORE_H_
+
+#include <botan/internal/asio_includes.h>
+#include <botan/tls_callbacks.h>
+#include <mutex>
+#include <vector>
+
+namespace Botan {
+
+namespace TLS {
+
+/**
+ * Contains the buffers for reading/sending, and the needed botan callbacks
+ */
+struct StreamCore : public Botan::TLS::Callbacks
+   {
+      struct Buffer
+         {
+         Buffer() : dynamicBuffer(data_buffer) {}
+         std::vector<uint8_t> data_buffer;
+         boost::asio::dynamic_vector_buffer<
+         uint8_t, typename decltype(data_buffer)::allocator_type>
+         dynamicBuffer;
+         };
+
+      StreamCore()
+         : m_input_buffer_space(17 * 1024, '\0'), // enough for a TLS Datagram
```

Comment:
17KB always? That's a lot...who am I to judge though.

Reply:
TLS maximum record size is 16K plus some overhead for headers and padding.

Reply:
When you say a "TLS Record" is that this thing:
https://github.com/vinniefalco/beast/blob/59347b3928a56b08b274221bb649e0fb60c229f4/include/boost/beast/core/detect_ssl.hpp#L108